File: experiments/anki_api_test.py
import json
import requests

def invoke(action, **params):
    """Gửi request tới AnkiConnect API"""
    requestJson = json.dumps({'action': action, 'params': params, 'version': 6})
    try:
        response = requests.post('http://localhost:8765', data=requestJson).json()
    except requests.exceptions.ConnectionError:
        print("Lỗi: Không thể kết nối tới Anki. Hãy chắc chắn rằng Anki đang mở và add-on AnkiConnect đã được cài đặt.")
        return None

    if len(response) != 2:
        raise Exception('response has an unexpected number of fields')
    if 'error' not in response:
        raise Exception('response is missing required error field')
    if 'result' not in response:
        raise Exception('response is missing required result field')
    if response['error'] is not None:
        raise Exception(response['error'])
    return response['result']

def test_anki_api():
    print("Bắt đầu test Anki API...\n")
    
    # 1. Lấy danh sách các decks
    decks = invoke('deckNames')
    if decks is None:
        return
        
    print("--- Danh sách các Decks ---")
    print(decks)
    
    if not decks:
        print("Không tìm thấy deck nào.")
        return
        
    # Chọn một deck để test (hoặc bạn có thể đổi tên deck ở đây)
    # Ví dụ: test_deck = "English"
    test_deck = decks[0] 
    if test_deck == "Default" and len(decks) > 1:
        test_deck = decks[1]
        
    print(f"\n=== Đang test trên deck: {test_deck} ===")
    
    # 2. Lấy các thẻ chưa học (is:new)
    # Cú pháp tìm kiếm Anki: deck:"Tên Deck" is:new
    new_cards_query = f'deck:"{test_deck}" is:new'
    new_cards = invoke('findCards', query=new_cards_query)
    
    print(f"\n--- Các thẻ CHƯA HỌC (mới) trong '{test_deck}' ---")
    print(f"Query: {new_cards_query}")
    print(f"Số lượng thẻ chưa học: {len(new_cards)}")
    if new_cards:
        print(f"ID của một số thẻ: {new_cards[:5]}{'...' if len(new_cards) > 5 else ''}")
        # Bạn có thể dùng action 'cardsInfo' để lấy nội dung chi tiết của các thẻ này
        # info = invoke('cardsInfo', cards=new_cards[:1])
        # print("Chi tiết thẻ đầu tiên:", json.dumps(info[0], indent=2, ensure_ascii=False))
    
    # 3. Lấy các thẻ cần ôn tập (is:due)
    # Cú pháp tìm kiếm Anki: deck:"Tên Deck" is:due
    due_cards_query = f'deck:"{test_deck}" is:due'
    due_cards = invoke('findCards', query=due_cards_query)
    
    print(f"\n--- Các thẻ CẦN ÔN (due) trong '{test_deck}' ---")
    print(f"Query: {due_cards_query}")
    print(f"Số lượng thẻ cần ôn: {len(due_cards)}")
    if due_cards:
        print(f"ID của một số thẻ: {due_cards[:5]}{'...' if len(due_cards) > 5 else ''}")
        
    # Lấy một vài ID thẻ (nếu có) để test chức năng lấy thông tin thẻ
    sample_card_ids = []
    if new_cards:
        sample_card_ids.extend(new_cards[:3])
    if due_cards and len(sample_card_ids) < 3:
        sample_card_ids.extend(due_cards[:3-len(sample_card_ids)])
        
    if sample_card_ids:
        # 4. Lấy nội dung của 1 thẻ
        card_id = sample_card_ids[0]
        print(f"\n--- Lấy nội dung của 1 thẻ (ID: {card_id}) ---")
        single_card_info = invoke('cardsInfo', cards=[card_id])
        if single_card_info:
            print("Thông tin chi tiết:")
            print(json.dumps(single_card_info[0], indent=2, ensure_ascii=False))
            
        # 5. Lấy nội dung của nhiều thẻ (n thẻ)
        n = len(sample_card_ids)
        if n > 1:
            print(f"\n--- Lấy nội dung của {n} thẻ (IDs: {sample_card_ids}) ---")
            multiple_cards_info = invoke('cardsInfo', cards=sample_card_ids)
            if multiple_cards_info:
                for i, info in enumerate(multiple_cards_info):
                    print(f"\n[Thẻ {i+1} - ID: {info['cardId']}] Các trường dữ liệu (fields):")
                    # In ra phần 'fields' để xem nội dung mặt trước/sau của thẻ
                    print(json.dumps(info['fields'], indent=2, ensure_ascii=False))
    else:
        print("\nKhông có thẻ mới hoặc thẻ cần ôn nào trong deck này để test lấy nội dung.")

if __name__ == '__main__':
    test_anki_api()

File: experiments/anki_api_test.js
async function invoke(action, params = {}) {
    try {
        // Node.js >= 18 có sẵn hàm fetch tích hợp
        const response = await fetch('http://localhost:8765', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({ action, version: 6, params })
        });
        
        const data = await response.json();
        
        if (Object.keys(data).length !== 2) {
            throw new Error('response has an unexpected number of fields');
        }
        if (!data.hasOwnProperty('error')) {
            throw new Error('response is missing required error field');
        }
        if (!data.hasOwnProperty('result')) {
            throw new Error('response is missing required result field');
        }
        if (data.error) {
            throw new Error(data.error);
        }
        
        return data.result;
    } catch (error) {
        // Bắt lỗi khi không thể kết nối tới cổng 8765 (Anki chưa mở hoặc chưa cài AnkiConnect)
        if (error.cause && error.cause.code === 'ECONNREFUSED') {
            console.error("Lỗi: Không thể kết nối tới Anki. Hãy chắc chắn rằng Anki đang mở và add-on AnkiConnect đã được cài đặt.");
            return null;
        }
        console.error("Lỗi:", error.message);
        return null;
    }
}

async function testAnkiApi() {
    console.log("Bắt đầu test Anki API bằng Node.js...\n");

    // 1. Lấy danh sách các decks
    const decks = await invoke('deckNames');
    if (!decks) return;

    console.log("--- Danh sách các Decks ---");
    console.log(decks);

    if (decks.length === 0) {
        console.log("Không tìm thấy deck nào.");
        return;
    }

    // Chọn một deck để test (hoặc bạn có thể đổi tên deck ở đây)
    let testDeck = decks[0];
    if (testDeck === "Default" && decks.length > 1) {
        testDeck = decks[1]; // Bỏ qua deck Default nếu có deck khác
    }

    console.log(`\n=== Đang test trên deck: ${testDeck} ===`);

    // 2. Lấy các thẻ chưa học (is:new)
    const newCardsQuery = `deck:"${testDeck}" is:new`;
    const newCards = await invoke('findCards', { query: newCardsQuery });
    
    console.log(`\n--- Các thẻ CHƯA HỌC (mới) trong '${testDeck}' ---`);
    console.log(`Query: ${newCardsQuery}`);
    console.log(`Số lượng thẻ chưa học: ${newCards ? newCards.length : 0}`);
    if (newCards && newCards.length > 0) {
        const sampleIds = newCards.slice(0, 5);
        console.log(`ID của một số thẻ: ${sampleIds.join(', ')}${newCards.length > 5 ? '...' : ''}`);
    }

    // 3. Lấy các thẻ cần ôn tập (is:due)
    const dueCardsQuery = `deck:"${testDeck}" is:due`;
    const dueCards = await invoke('findCards', { query: dueCardsQuery });
    
    console.log(`\n--- Các thẻ CẦN ÔN (due) trong '${testDeck}' ---`);
    console.log(`Query: ${dueCardsQuery}`);
    console.log(`Số lượng thẻ cần ôn: ${dueCards ? dueCards.length : 0}`);
    if (dueCards && dueCards.length > 0) {
        const sampleIds = dueCards.slice(0, 5);
        console.log(`ID của một số thẻ: ${sampleIds.join(', ')}${dueCards.length > 5 ? '...' : ''}`);
    }

    // Gộp ID để test lấy thông tin
    let sampleCardIds = [];
    if (newCards && newCards.length > 0) {
        sampleCardIds.push(...newCards.slice(0, 3));
    }
    if (dueCards && dueCards.length > 0 && sampleCardIds.length < 3) {
        sampleCardIds.push(...dueCards.slice(0, 3 - sampleCardIds.length));
    }

    if (sampleCardIds.length > 0) {
        // 4. Lấy nội dung của 1 thẻ
        const cardId = sampleCardIds[0];
        console.log(`\n--- Lấy nội dung của 1 thẻ (ID: ${cardId}) ---`);
        const singleCardInfo = await invoke('cardsInfo', { cards: [cardId] });
        if (singleCardInfo && singleCardInfo.length > 0) {
            console.log("Thông tin chi tiết:");
            console.log(JSON.stringify(singleCardInfo[0], null, 2));
        }

        // 5. Lấy nội dung của nhiều thẻ (n thẻ)
        const n = sampleCardIds.length;
        if (n > 1) {
            console.log(`\n--- Lấy nội dung của ${n} thẻ (IDs: ${sampleCardIds.join(', ')}) ---`);
            const multipleCardsInfo = await invoke('cardsInfo', { cards: sampleCardIds });
            if (multipleCardsInfo) {
                multipleCardsInfo.forEach((info, i) => {
                    console.log(`\n[Thẻ ${i + 1} - ID: ${info.cardId}] Các trường dữ liệu (fields):`);
                    console.log(JSON.stringify(info.fields, null, 2));
                });
            }
        }
    } else {
        console.log("\nKhông có thẻ mới hoặc thẻ cần ôn nào trong deck này để test lấy nội dung.");
    }
}

testAnkiApi();
